File: server/controllers/pathController.js
const db = require('../config/firebase');

const pathController = {

    // Get all paths
    getAllPaths: async (req, res) => {
        try {
            const pathsSnapshot = await db.ref('/paths').get();
            const paths = pathsSnapshot.val() || {}; // Return empty object if no paths found
            res.status(200).json(paths);
        } catch (error) {
            console.error('Error getting all paths', error);
            res.status(500).json({ error: error.message });
        }
    },

    // Add new path
    addPath: async (req, res) => {
        try {
            const { initialTPS, endTPS, distance } = req.body;

            // Check if both initialTPS and endTPS exist in the TPS collection
            const initialTPSSnapshot = await db.ref(`/tps/${initialTPS}`).get();
            const endTPSSnapshot = await db.ref(`/tps/${endTPS}`).get();

            if (!initialTPSSnapshot.exists() || !endTPSSnapshot.exists()) {
                return res.status(404).json({ message: 'One or both TPS locations do not exist' });
            }

            // Create a new path object
            const newPath = {
                initialTPS,  // Foreign key to initial TPS
                endTPS,      // Foreign key to end TPS
                distance     // Distance between the TPS locations
            };

            // Save the new path to Firebase
            const newPathRef = await db.ref('/paths').push(newPath);
            res.status(201).json({ message: 'Path added successfully', path: { ...newPath, pathId: newPathRef.key } });
        } catch (error) {
            console.error('Error adding path', error);
            res.status(500).json({ error: error.message });
        }
    },

    // Get path by ID
    getPathById: async (req, res) => {
        try {
            const pathId = req.params.pathId;

            if (!pathId || typeof pathId !== 'string') {
                return res.status(400).json({ error: 'Invalid path ID' });
            }

            const pathsnapshot = await db.ref(`/paths/${pathId}`).get();

            if (!pathsnapshot.exists()) {
                return res.status(404).json({ message: 'Path not found' });
            }

            const path = pathsnapshot.val();
            res.status(200).json(path);
        } catch (error) {
            console.error('Error getting path by ID', error);
            res.status(500).json({ error: error.message });
        }
    },

    // Update path by ID
    updatePath: async (req, res) => {
        try {
            const pathId = req.params.pathId;

            if (!pathId || typeof pathId !== 'string') {
                return res.status(400).json({ error: 'Invalid path ID' });
            }
            const { initialTPS, endTPS, distance } = req.body;

            // Check if path exists
            const pathsnapshot = await db.ref(`/paths/${pathId}`).get();
            if (!pathsnapshot.exists()) {
                return res.status(404).json({ message: 'path not found' });
            }

            // Update the path with new data
            const updatedPath = {
                initialTPS,
                endTPS,
                distance
            };

            await db.ref(`/paths/${pathId}`).update(updatedPath);
            res.status(200).json({ message: 'Path updated successfully', path: updatedPath });
        } catch (error) {
            console.error('Error updating path', error);
            res.status(500).json({ error: error.message });
        }
    },

    // Delete path
    deletePath: async (req, res) => {
        try {
            const pathId = req.params.pathId;

            if (!pathId || typeof pathId !== 'string') {
                return res.status(400).json({ error: 'Invalid path ID' });
            }

            // Check if path exists
            const pathsnapshot = await db.ref(`/paths/${pathId}`).get();
            if (!pathsnapshot.exists()) {
                return res.status(404).json({ message: 'Path not found' });
            }

            // Delete the path
            await db.ref(`/paths/${pathId}`).remove();
            res.status(200).json({ message: 'Path deleted successfully' });
        } catch (error) {
            console.error('Error deleting path', error);
            res.status(500).json({ error: error.message });
        }
    }
};

module.exports = pathController;

File: backend/app/controllers/path_controller.py
from flask import Blueprint, request, jsonify
from firebase_admin import db

path_controller = Blueprint('path_controller', __name__)

# Get all paths
@path_controller.route('/paths', methods=['GET'])
def get_all_paths():
    try:
        paths_snapshot = db.reference('/paths').get()
        paths = paths_snapshot or {}  # Return empty dictionary if no paths found
        return jsonify(paths), 200
    except Exception as error:
        print('Error getting all paths:', error)
        return jsonify({'error': str(error)}), 500

# Add new path
@path_controller.route('/paths', methods=['POST'])
def add_path():
    try:
        data = request.get_json()
        path_name = data.get('pathName')
        initial_tps = data.get('initialTPS')
        end_tps = data.get('endTPS')
        distance = data.get('distance')

        # Check if both initialTPS and endTPS exist in the TPS collection
        initial_tps_snapshot = db.reference(f'/tps/{initial_tps}').get()
        end_tps_snapshot = db.reference(f'/tps/{end_tps}').get()

        if not initial_tps_snapshot or not end_tps_snapshot:
            return jsonify({'message': 'One or both TPS locations do not exist'}), 404

        # Create a new path object
        new_path = {
            'pathName': path_name,
            'initialTPS': initial_tps,
            'endTPS': end_tps,
            'distance': distance
        }

        # Save the new path to Firebase
        new_path_ref = db.reference('/paths').push(new_path)
        return jsonify({'message': 'Path added successfully', 'path': {**new_path, 'pathId': new_path_ref.key}}), 201
    except Exception as error:
        print('Error adding path:', error)
        return jsonify({'error': str(error)}), 500

# Get path by ID
@path_controller.route('/paths/<path_id>', methods=['GET'])
def get_path_by_id(path_id):
    try:
        if not path_id or not isinstance(path_id, str):
            return jsonify({'error': 'Invalid path ID'}), 400

        path_snapshot = db.reference(f'/paths/{path_id}').get()

        if not path_snapshot:
            return jsonify({'message': 'Path not found'}), 404

        return jsonify(path_snapshot), 200
    except Exception as error:
        print('Error getting path by ID:', error)
        return jsonify({'error': str(error)}), 500

# Update path by ID
@path_controller.route('/paths/<path_id>', methods=['PUT'])
def update_path(path_id):
    try:
        if not path_id or not isinstance(path_id, str):
            return jsonify({'error': 'Invalid path ID'}), 400

        data = request.get_json()
        initial_tps = data.get('initialTPS')
        end_tps = data.get('endTPS')
        distance = data.get('distance')

        # Check if path exists
        path_snapshot = db.reference(f'/paths/{path_id}').get()
        if not path_snapshot:
            return jsonify({'message': 'Path not found'}), 404

        # Update the path with new data
        updated_path = {
            'initialTPS': initial_tps,
            'endTPS': end_tps,
            'distance': distance
        }

        db.reference(f'/paths/{path_id}').update(updated_path)
        return jsonify({'message': 'Path updated successfully', 'path': updated_path}), 200
    except Exception as error:
        print('Error updating path:', error)
        return jsonify({'error': str(error)}), 500

# Delete path by ID
@path_controller.route('/paths/<path_id>', methods=['DELETE'])
def delete_path(path_id):
    try:
        if not path_id or not isinstance(path_id, str):
            return jsonify({'error': 'Invalid path ID'}), 400

        # Check if path exists
        path_snapshot = db.reference(f'/paths/{path_id}').get()
        if not path_snapshot:
            return jsonify({'message': 'Path not found'}), 404

        # Delete the path
        db.reference(f'/paths/{path_id}').delete()
        return jsonify({'message': 'Path deleted successfully'}), 200
    except Exception as error:
        print('Error deleting path:', error)
        return jsonify({'error': str(error)}), 500
